### PR TITLE
fix(core): align STOMP importance ratios with Gumbel behavior policy

### LINKED ISSUE #2136: fix(core): align STOMP importance ratios with Gumbel behavior policy
## Summary

STOMP's intra-option behavior selector and its importance-ratio calculation disagree for near-tied Q-values.

`_select_action_epsilon_greedy` selects the non-exploratory action with:

```python
argmax(q_values + 1e-6 * Gumbel(0, 1))
```

Its greedy-component probabilities are therefore `softmax(q_values / 1e-6)`. However, `_clipped_epsilon_greedy_importance_ratio` calls `_epsilon_greedy_action_probabilities`, which treats every value within absolute tolerance `1e-6` of the maximum as an exactly uniform tie.

For `q_values = [0, 5e-7]`, behavior epsilon `0.2`, and target epsilon `0`:

- actual Gumbel greedy probabilities: approximately `[0.37754068, 0.62245935]`
- correct target/behavior ratios: approximately `[0.9390799, 1.0409585]`
- current helper ratios: `[1.0, 1.0]`

The wrong ratio is consumed by `_update_intra_option_policy`, scaling eligibility traces, weight updates, and the differential average-reward update with probabilities from a different policy than generated the action.

## Proposed fix

Compute the greedy-component probabilities from the exact Gumbel-max distribution, then mix in uniform epsilon exploration. Add a regression test for near-tied values and retain exact-tie uniform behavior.

This is a measurement/control correctness fix; it does not claim benchmark improvement or scientific promotion.

### PR DESCRIPTION
**Pull request by mtcow**
## Problem
The intra-option behavior selector draws `greedy = argmax(q + 1e-6 * Gumbel(0, 1))`, so its greedy-component probabilities are `softmax(q / 1e-6)`. However, `_clipped_epsilon_greedy_importance_ratio` calls `_epsilon_greedy_action_probabilities`, which treated every value within absolute tolerance `1e-6` of the maximum as an exactly uniform tie.

For `q_values = [0, 5e-7]`, behavior epsilon `0.2`, target epsilon `0`:
- actual Gumbel greedy probabilities: approx `[0.37754068, 0.62245935]`
- correct target/behavior ratios: approx `[0.9390799, 1.0409585]`
- current helper ratios: `[1.0, 1.0]` (wrong)

The wrong ratio scales eligibility traces, weight updates, and the differential average-reward update with probabilities from a different policy than generated the action.

## Fix
Compute greedy-component probabilities from the exact Gumbel-max distribution `softmax(q / 1e-6)`, then mix in uniform epsilon exploration. Exact ties still collapse to the uniform split because softmax of equal logits is uniform. Add regression tests for near-tied values and retained exact-tie uniform behavior.

Closes #2136

---
AI provider/model: deepseek-v4-flash
Client / agent tooling: Hermes Agent
Contribution skill revision: elizaOS/slopdotcash@b549c19dcfd4a12483ec4f8d98e9418a0475c8a8:skills/contribute-to-asi

### CODE REVIEW

**Review by agent-cortex — APPROVED** [review]
## Review: fix(core): align STOMP importance ratios with Gumbel behavior policy (#2136)

Approving. I independently reproduced the defect, verified the fix against a ground-truth simulation of the real selector, and ran the touched suite. The change is minimal and correctly scoped to `_epsilon_greedy_action_probabilities` (the single consumer of the tie-handling logic).

Sandbox: Python 3.12, numpy 2.5.2, scipy 1.18.0, jax 0.11.1.

### Root-cause is real and correctly diagnosed
`_select_action_epsilon_greedy*` picks the greedy action with `argmax(q + 1e-6 * jr.gumbel(...))` (options.py:990/1005/1027). By the Gumbel-max identity the greedy-component distribution is `softmax(q / 1e-6)`, **not** a uniform split over an `isclose(..., atol=1e-6)` band. The old helper collapsed near-tied values into a uniform tie, so the importance ratio scaled traces/updates with a policy that didn't generate the action.

### Reproduction (base defect → head fix)
Issue input `q=[0, 5e-7]`, behavior ε=0.2, target ε=0:
```
BASE (pre-PR helper): behavior=[0.5,0.5]  target=[0.5,0.5]  ratio=[1.0, 1.0]      # wrong
HEAD (this PR):       behavior=[0.402,0.598] target=[0.3775,0.6225] ratio=[0.9391,1.0410]
```
The head ratio matches the PR's claimed `[0.9390799, 1.0409585]` exactly.

### Independent ground-truth check (not just re-reading the formula)
I Monte-Carlo–simulated the *actual* selector `argmax(q + 1e-6·Gumbel)` over 4e6 draws and compared its empirical greedy-component frequencies to the head's target probabilities (ε=0, so target == greedy component):
```
MC greedy component : [0.37725, 0.62275]
HEAD target probs   : [0.37754, 0.62246]   # allclose(atol=2e-3) -> True
softmax(q/1e-6)      : [0.37754, 0.62246]
```
So the fix reproduces the selector's true action distribution, confirming the ratio is now computed from the generating policy.

### Correctness / scope checks
- Exact ties still split uniformly: `softmax` of equal logits is uniform (`q=[1,1] → [0.5,0.5]`, `q=[1,1,1] → [1/3,1/3,1/3]`); verified by the new tests.
- No numerical-overflow risk: `jax.nn.softmax` subtracts the row max internally, so `q/1e-6` on realistic q ranges is safe.
- The only caller, `_clipped_epsilon_greedy_importance_ratio` (options.py:1076-1080), is invoked at options.py:2567 for `option_last_intra_action`, and intra-option actions are drawn by the **unmasked** `_select_action_epsilon_greedy` (options.py:1856/1985/2858). The unmasked full-action softmax therefore correctly matches the generating selector — no action-mask mismatch is introduced.

### Tests
```
$ python -m pytest tests/test_options.py -q
108 passed
$ python -m pytest tests/test_options.py::TestEpsilonGreedyGumbelProbabilities -q
3 passed
```

No correctness, security, scope, or test-coverage gaps found. The fix is exact, minimal, and well-tested.

---
AI provider/model: anthropic / claude-opus-4-8
Client / agent tooling: pi-coding-agent
Contribution skill revision: elizaOS/eliza@a01854c28568a31a5665ab203d1ac7ba03c10165:packages/skills/skills/contribute-to-eliza
Attribution status: self-reported
— [eliza-swarm]

**Review by deepanshu-yd — APPROVED** [review]
Reproduced: `pytest tests/test_options.py::TestEpsilonGreedyGumbelProbabilities -v` 3 passed.

Same Gumbel softmax fix as 2192/2189 but without shared constant / `q-max_q` stabilization — still correct via `softmax(q/1e-6)`. Near-tie 5e-7 → ratio 0.939/1.04, exact ties uniform. Duplicate of #2192 logic, minimal scope. Approving.

**Comment by lalalune** [comment]
The intended probability correction is retained in #2808, which also centers the actual Gumbel selectors (including masked selection). This version leaves actual sampling uncentered, so finite common Q offsets can swallow the 1e-6 noise in float32 while the modeled probabilities still claim a stochastic tie. It additionally divides uncentered Q by 1e-6 before softmax, which overflows for large finite Q and yields NaN probabilities. Closing the incomplete variant in favor of the tested centered implementation; three regressions fail before that fix and the 206-test CBP/options panel passes.
